test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java

```diff
@@ -0,0 +1,1044 @@
+ * for IR matching).
+ * <p>
+ * After the flag VM terminates, it starts a new <i>test VM</i> which performs the execution of the specified
+ * tests in the test class as described in {@link Test}, {@link Check}, and {@link Run}.
+ * <p>
+ * In a last step, once the test VM has terminated without exceptions, IR matching is performed if there are any IR
+ * rules and if no VM flags disable it (e.g. not running with {@code -Xint}, see {@link IR} for more details).
+ * The IR regex matching is done on the output of {@code -XX:+PrintIdeal} and {@code -XX:+PrintOptoAssembly} by parsing
+ * the hotspot_pid file of the test VM. Failing IR rules are reported by throwing a {@link IRViolationException}.
+ *
+ * @see Test
+ * @see Check
+ * @see Run
+ * @see IR
+ */
+public class TestFramework {
+    /**
+     * JTreg can define additional VM (-Dtest.vm.opts) and Javaoptions (-Dtest.java.opts) flags. IR verification is only
+     * performed when all these additional JTreg flags (does not include additionally added framework and scenario flags
+     * by user code) are whitelisted.
+     *
+     * A flag is whitelisted if it is a property flag (starting with -D), -ea, -esa, or if the flag name contains any of
+     * the entries of this list as a substring (partial match).
+     */
+    public static final Set<String> JTREG_WHITELIST_FLAGS = new HashSet<>(
+            Arrays.asList(
+                    // The following substrings are part of more than one VM flag
+                    "RAM",
+                    "Heap",
+                    "Trace",
+                    "Print",
+                    "Verify",
+                    "TLAB",
+                    "UseNewCode",
+                    // The following substrings are only part of one VM flag (=exact match)
+                    "CreateCoredumpOnCrash",
+                    "IgnoreUnrecognizedVMOptions",
+                    "UnlockDiagnosticVMOptions",
+                    "UnlockExperimentalVMOptions",
+                    "BackgroundCompilation",
+                    "Xbatch",
+                    "TieredCompilation"
+            )
+    );
+
+    static final boolean VERBOSE = Boolean.getBoolean("Verbose");
+    static final boolean TESTLIST = !System.getProperty("Test", "").isEmpty();
+    static final boolean EXCLUDELIST = !System.getProperty("Exclude", "").isEmpty();
+    static final String TEST_VM_FLAGS_START = "##### TestFrameworkPrepareFlags - used by TestFramework #####";
+    static final String TEST_VM_FLAGS_DELIMITER = " ";
+    static final String TEST_VM_FLAGS_END = "----- END -----";
+    static final String RERUN_HINT = """
+                                       #############################################################
+                                        - To only run the failed tests use -DTest, -DExclude,
+                                          and/or -DScenarios.
+                                        - To also get the standard output of the test VM run with\s
```

Comment:
why do you need `\s` here?

Reply:
I needed this because `jcheck` complained about trailing whitespaces.

Reply:
well, why do you want to have a trailing space here? the text block you current have produces the following multeline sting, which has a trailing space at 4th line:
```  #############################################################
   - To only run the failed tests use -DTest, -DExclude,
     and/or -DScenarios.
   - To also get the standard output of the test VM run with 
     -DReportStdout=true or for even more fine-grained logging
     use -DVerbose=true.
  #############################################################```

Reply:
Hm, it somehow made sense when I added it back there but looking at it again it no longer does. I'll remove it in the next iteration.